### PR TITLE
PLT-8302 - Fixing gifv files not displaying correctly

### PR DESCRIPTION
**Pull request by jkurian**
### Summary

Renders .gifv urls to a video html element as opposed to previously where they were rendered into an  html element.

Had a lot of trouble squashing last two commits so I am leaving it as of now. I want to submit the PR now because I would like feedback.

### Directories/Files Changed (All in mattermost-webapp)

**`components/post_view/post_mp4`**: Created a new component which will display the gifv link. Since it is a video, we display it in a <video> HTML element. Added error handling for video by created an HTML5 video element, and then adding the onerror and onload functions to the video which will be called appropriated after video.load() has been executed. gifv links now appear in a preview in the chat. Have to apply this rendering to \ next.

**`utils/markdown/renderer.jsx`**: Had to add functionality to image(href, title, text) function. Basically the mark package lacks ability to render a video element and just detects for images when user tries to unfurl an image with \. So instead of changing the mark package to include a video() function I applied to string regex check to see if the user is trying to post a gifv file. If there is a match, then we render a <video> HTML element rather then an image element which will not display gifv as wanted.

**`components/post_view/post_body_additional_content/post_body_additional_content.jsx`**: Added a isLinkGifv(link) function which checks if the URL is a gifv address. I used the same regex matching used for determining if the URL is a gif, jpg, bmp etc. If the link is a gifv, I render the PostMp4 component and return it.

**`utils/post_utils.jsx`**: Added a changeMp4Src and a changeJPGSrc function which takes in a src URL and returns the URL with the gifv extension replaced with .mp4 or jpg which are used in the <video> HTML element.

### Ticket Link
PLT-8302 - https://github.com/mattermost/mattermost-server/issues/8302

### Checklist
**Apologies for any items I missed in the checklist, please let me know if I need to include more!**
- [X] Ran `make check-style` to check for style errors (required for all pull requests)
- [X] Ran `make test` to ensure unit and component tests passed
- [X] Touches critical sections of the codebase (auth, posting, etc.)

### CODE REVIEW

**Comment by amyblais** [comment]
Hi @jkurian Thank you for this contribution! 
I tested this and it wasn't working for me - I posted  in a channel and a picture didn't appear - perhaps I misunderstood something? I'm happy to help re-test if you have any suggestions or if there is another image that I can use for testing!

**Comment by jkurian** [comment]
Hi @amyblais!

It will actually only render a video if the URL is a valid gifv resource. For example, if you try ```, it should show a video file in the chat. Let me know if this works for you!

**Comment by amyblais** [comment]
Hi @jkurian,

`` works well on webapp!

I just have a few questions:
1. When I post the link (` on mobile app, the picture appears in the webapp but not on mobile - is this expected?
2. Also, in the Test User Account there is lot of white space between the LHS and the text box / centre pane - is this part of the feature?

**Comment by jkurian** [comment]
Hi again @amyblais!

Thanks for your feedback! 

I did some reading on the mobile app video rendering and found out is should be a simple fix. Basically the way the video HTML element is set up isnt supported on mobile, so we would have to render it another way. I have a few ideas which I will try to get to soon!

I will also have to take a closer look at the white space and will update you when I get a chance.

**Comment by amyblais** [comment]
@jkurian,

1. No worries, the Jira ticket also mentioned mobile app so I thought it might be part of the fix - feel free to include it in this PR or open another PR for the mobile fix, whichever is easiest.
2. Here is a screenshot - I added the red line to show where I see the extra white space - I'm not sure whether it's part of the feature or not, perhaps it's just what I see on my screen:
screenshot_59

**Comment by amyblais** [comment]
@jkurian You don't need to worry about the extra white space, I noticed that the Channel Display Mode account setting was just set to "Fixed width, centered" instead of "Full width", so it's not an issue - sorry about that!

**Comment by jkurian** [comment]
Thanks for the feedback @esethna. I am going to be looking over this starting tomorrow and make the appropriate changes!

**Comment by esethna** [comment]
Hi @jkurian, just checking in, anything I can help with?

**Comment by jkurian** [comment]
hi @esethna,

Sorry for not being active here. I am getting pretty consumed by other work. 

I am still alive though and will look to get back to this as soon as I find some spare time (things should slow down for me soon).

**Comment by esethna** [comment]
Thanks @jkurian, let us know if you need any help with anything

**Comment by jkurian** [comment]
Hi @esethna, 

I have updated the code to make the embedded video element resize dynamically. Let me know how it looks on your end!

Jerry

**Comment by jkurian** [comment]
@esethna  Just noticed some extra whitespace beside the video element, I am updating the code now.

**Comment by esethna** [comment]
Thanks @jkurian! Looks great, moving to dev review

**Comment by esethna** [comment]
@jkurian would you be interested in submitting a PR to the [mobile repo](https://github.com/mattermost/mattermost-mobile) for this as well?

**Comment by jkurian** [comment]
@esethna Great!Yes, I could do that as well after the dev review for this PR.

**Comment by jkurian** [comment]
@esethna Great! Yes, I plan on doing that as well after the dev review for this PR. Is that ok?

**Comment by esethna** [comment]
@jkurian for sure, thanks!

File: utils/markdown/renderer.jsx [hunk 1]
```diff
@@ -120,19 +119,43 @@ export default class Renderer extends marked.Renderer {
...
-        if (dimensions.length > 0) {
-            out += ' width="' + dimensions[0] + '"';
-        }
-        if (dimensions.length > 1) {
-            out += ' height="' + dimensions[1] + '"';
+
+        //After looking into the mark package, there appears to be no in-line level render function in marked.js
+        //to render a video. Since we just want to render gifv files like we do gifs, we can check if
+        //the URL src is a gifv, then render it in a <video> HTML5 element rather than an <img> element without
+        //having to touch the mark package. There is some DRY issues here, but I am not sure how to overcome that.
```
Comment (saturninoabril) [inline utils/markdown/renderer.jsx]: Please update into a concise comment. Might be good to move above comments to PR description.

File: components/post_view/post_mp4/post_mp4.jsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                            this.props.link,
+                            this.props.hasImageProxy
+                        )}
+                        type='video/mp4'
+                    />
+                </video>
+            </div>
+        );
+    }
+}
```
Comment (saturninoabril) [inline components/post_view/post_mp4/post_mp4.jsx]: Please add component test.

File: components/post_view/post_mp4/post_mp4.jsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import {postListScrollChange} from 'actions/global_actions.jsx';
+import * as PostUtils from 'utils/post_utils.jsx';
```
Comment (saturninoabril) [inline components/post_view/post_mp4/post_mp4.jsx]: Please explicitly import only those that are needed by this component, like:
```
import {changeToJPGSrc , changeToMp4Src} as PostUtils from 'utils/post_utils.jsx';
```

File: utils/post_utils.jsx [hunk 4]
```diff
@@ -53,6 +53,22 @@ export function getImageSrc(src, hasImageProxy) {
...
+    return mp4Ext;
+}
+
+export function changeToJPGSrc(src, hasImageProxy) {
+    if (hasImageProxy) {
+        return Client4.getBaseRoute() + '/image?url=' + encodeURIComponent(src);
+    }
+    const jpgExt = src.substr(0, src.lastIndexOf('.gifv')) + '.jpg';
+    return jpgExt;
+}
```
[the remark this answers is not in the document]
Comment (jkurian) [reply utils/post_utils.jsx]: For the unit tests, I can just add to the test spec found at `/Users/JerryKurian/go/src/github.com/mattermost/mattermost-webapp/tests/utils/post_utils.test.jsx` correct?

[the remark this answers is not in the document]
Comment (saturninoabril) [reply utils/post_utils.jsx]: Yes

File: components/post_view/post_mp4/index.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply components/post_view/post_mp4/index.js]: Updated my comment

File: components/post_view/post_mp4/post_mp4.jsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            loaded: true,
+        });
+        if (this.props.onLinkLoadError) {
+            this.props.onLinkLoadError();
+        }
+    }
+
+    onImageClick = (e) => {
+        e.preventDefault();
+        this.props.handleImageClick();
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply components/post_view/post_mp4/post_mp4.jsx]: Should be like:
```
    onImageClick = (e) => {
        e.preventDefault();
        if(this.props.handleImageClick) {
            this.props.handleImageClick();
        }
    };
```

[the remark this answers is not in the document]
Comment (jkurian) [reply components/post_view/post_mp4/post_mp4.jsx]: Ahh ok, that makes sense. Thanks!

**Comment by esethna** [comment]
@saturninoabril yes, good catch, that would be a regression. @jkurian please help take a look when you get a chance?

**Comment by jkurian** [comment]
@esethna Will do!

File: utils/post_utils.jsx [hunk 7]
```diff
@@ -53,6 +53,28 @@ export function getImageSrc(src, hasImageProxy) {
     return src;
 }
 
+export function changeToMp4Src(src, hasImageProxy) {
```
Comment (hmhealey) [inline utils/post_utils.jsx]: Shouldn't these links be changed to mp4 and jpg before being passed to the image proxy?

File: utils/markdown/renderer.jsx [hunk 8]
```diff
@@ -120,19 +119,40 @@ export default class Renderer extends marked.Renderer {
...
-            out += ' height="' + dimensions[1] + '"';
+
+        const regex = /.+\/(.+\.(gifv))(?:\?.*)?$/i;
+        const match = src.match(regex);
+        let out = '';
+
+        //If the URL src is a gifv, then render it in a <video> HTML5 element, otherwise it will go into an <img> element
+        if (match) {
+            const jpgSrc = src.substr(0, src.lastIndexOf('.gifv')) + '.jpg';
+            const mp4Src = src.substr(0, src.lastIndexOf('.gifv')) + '.mp4';
```
Comment (hmhealey) [inline utils/markdown/renderer.jsx]: Is it necessary to change the extension on the video source? I think `<video><source src="https://i.imgur.com/abc123.gifv"/></video>` might be valid

File: utils/markdown/renderer.jsx [hunk 9]
```diff
@@ -120,19 +119,40 @@ export default class Renderer extends marked.Renderer {
...
-        }
-        if (dimensions.length > 1) {
-            out += ' height="' + dimensions[1] + '"';
+
+        const regex = /.+\/(.+\.(gifv))(?:\?.*)?$/i;
+        const match = src.match(regex);
+        let out = '';
+
+        //If the URL src is a gifv, then render it in a <video> HTML5 element, otherwise it will go into an <img> element
+        if (match) {
```
Comment (hmhealey) [inline utils/markdown/renderer.jsx]: We'll want to use the image proxy to load these, if possible. Have you tried testing this with the image proxy? If not, we'll want to do that

**Comment by amyblais** [comment]
@jkurian I'm not able to repro the bug that Saturn reported - @saturninoabril can you help give more details?

**Comment by saturninoabril** [comment]
Thanks @amyblais for checking. Not sure what happened but you're right.
@jkurian Please disregard the issue/regression that I reported earlier.

**Comment by jkurian** [comment]
@saturninoabril Great! It might have been the account user settings. If you log in as a regular user, the default is set to display links as `collapsed` rather than `expanded` so only the link url will show up and no preview. For the sysadmin account, the default is set to `expanded` and you would be able to see the preview.

File: utils/markdown/renderer.jsx (hunk 8 above)
Comment (jkurian) [reply utils/markdown/renderer.jsx]: I will look into this soon!

File: components/post_view/post_mp4/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkurian) [reply components/post_view/post_mp4/index.js]: It looks like the linter settings were changed and I get a lint error with any header that is not 
`// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.`

File: tests/components/post_view/post_mp4.test.jsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        type: '',
+        update_at: 1502715372443,
+        user_id: 'b4pfxi8sn78y8yq7phzxxfor7h',
+    };
+
+    test('should match snapshot', () => {
+        const wrapper = shallow(<PostMp4 {...post}/>);
+        expect(wrapper).toMatchSnapshot();
+    });
+});
```
[the remark this answers is not in the document]
Comment (jkurian) [reply tests/components/post_view/post_mp4.test.jsx]: How does it look now?

File: utils/markdown/renderer.jsx (hunk 9 above)
Comment (jkurian) [reply utils/markdown/renderer.jsx]: Hmm no, I will look into this as well.

Comment (jkurian) [reply utils/markdown/renderer.jsx]: I added some tests with basically the hasImageProxy boolean set to true. Is this ok?

File: utils/post_utils.jsx (hunk 7 above)
Comment (jkurian) [reply utils/post_utils.jsx]: I made the suggested changes! Let me know if it is ok :)

**Comment by jkurian** [comment]
Yes, looks like I was using yarn. I removed the yarn.lock 👍

**Comment by hmhealey** [comment]
Code looks good, but I'd like to test this tomorrow with a server with an image proxy to see how that works.

**Comment by jkurian** [comment]
@hmhealey Thanks for taking the time to test this. To understand better, the idea is that the image proxy allows for local servers to store the image data, preventing the need to go fetch the data from a public / external sever? The problem here is that the gifv files can't be stored via image proxy, and thus will always fetch data from the external servers when it needs to be rendered on the page?

So what would need to happen is allowing for video files stored via the image proxy after being posted to prevent the browser from fetching the data from external servers when the page with the video file is rendered?

**Comment by hmhealey** [comment]
That's correct. Unfortunately, the image proxy that we use ([link](https://github.com/atmos/camo)) doesn't support storing videos at all. Since we provide that proxy functionality as a promise that the browser won't connect to any insecure 3rd party sites, we can't accept this PR in its current form since it breaks that promise.

**Comment by jkurian** [comment]
Ahh I understand, thats a shame. Is there anything I can do from here? 

If not, we can close this PR and maybe come back to it in the future if we find a workaround while I try to tackle another issue!

**Comment by cpanato** [comment]
Removed the test server.

**Comment by esethna** [comment]
Thanks @jkurian, added the work in progress label until the proxy changes are complete.

**Comment by shermdog** [comment]
@hmhealey @esethna  I've had success using go-camo which supports proxy of video elements  

https://github.com/cactus/go-camo/commit/a92ce5e48744df976930bba368089a6d39f58cd7

What are your thoughts on getting this re-opened and reviewed against go-camo?

**Comment by hmhealey** [comment]
I'd certainly like to, but there's a few big changes in the works right now that'll affect this heavily, so it may be best to let the dust settle around those before resuming work on this.

The first change is that we're making the client aware of the dimensions of all content whenever it receives posts so that the layout doesn't jump around as everything loads in. That would require some additional work on the server to send those down to the client like it will do with images.

The second change is to add a built-in image proxy to the server. I'm planning on using [willnorris/imageproxy](https://github.com/willnorris/imageproxy) which also supports videos, but that will require some more changes that will affect this part of the code as well

**Pull request closed without merging**